File: 2018/aoc_day_02/src/main.rs
use std::collections::HashMap;
use aoc_util::aoc_util::*;
use std::time::Instant;

fn main() -> AocResult<()> {
    let problem_input = get_problem_input()?;
    println!("Checksum: {}", part_1(&problem_input)?);
    let start = Instant::now();
    println!("A: ID differing by one character, common chars only: {}", part_2a(&problem_input)?);
    let after_a = Instant::now();
    println!("B: ID differing by one character, common chars only: {}", part_2b(&problem_input)?);
    let after_b = Instant::now();

    println!("2a: {:?}  2b: {:?}", after_a - start, after_b - after_a);
    Ok(())
}

/// Executes logic for part 1 of Advent of Code 2018 day 2.
fn part_1(input: &String) -> AocResult<i32> {
    let mut three_count = 0;
    let mut two_count = 0;

    for ln in input.lines() {
        let mut letters_count: HashMap<char,u32> = HashMap::new();
        ln.chars().for_each(|c| *letters_count.entry(c).or_insert(0) += 1);
        if letters_count.iter().any(|l| *l.1 == 3) {
            three_count += 1;
        }
        if letters_count.iter().any(|l| *l.1 == 2) {
            two_count += 1;
        }
    }

    Ok(three_count * two_count)
}

fn part_2a(input: &String) -> AocResult<String> {
    for ln_a in input.lines() {
        for ln_b in input.lines().skip(1) {
            if ln_a.len() != ln_b.len() {
                continue;
            }

            let mut common_c = String::with_capacity(ln_a.len());
            ln_a.chars().zip(ln_b.chars()).filter(|(a,b)| a==b).for_each(|(a,_b)| common_c.push(a));
            if common_c.len() == ln_a.len() - 1 {
                return Ok(common_c);
            }
        }
    }

    Err("Nothing found.".into())
}

fn part_2b(input: &String) -> AocResult<String> {
    for ln_a in input.lines() {
        for ln_b in input.lines().skip(1) {
            if ln_a.len() != ln_b.len() {
                continue;
            }

            let mut common_c = String::with_capacity(ln_a.len());
            let mut diff_count = 0;
            for (a, b) in ln_a.chars().zip(ln_b.chars()){
                if a != b {
                    diff_count += 1;
                    if diff_count > 1 {
                        break;
                    }
                }
                else {
                    common_c.push(a);
                }
            }

            if diff_count == 1 {
                return Ok(common_c);
            }
        }
    }

    Err("Nothing found.".into())
}

File: 2018/aoc_day_02/src/main.py

import time
import sys


def main(inputfile):
    with open(inputfile) as f:
        input_str = f.read()

        start = time.perf_counter() * 1000
        print(f"2a: {part_2a(input_str)}")
        after_a = time.perf_counter() * 1000
        print(f"2b: {part_2b(input_str)}")
        after_b = time.perf_counter() * 1000

        print(f"{after_a - start}, {after_b - after_a}")


def part_2a(input_str):
    for ln_a in input_str.splitlines():
        for ln_b in input_str.splitlines()[1:]:
            if len(ln_a) != len(ln_b):
                continue
            common_c = ""
            for c in filter(lambda t: t[0] == t[1], zip(ln_a, ln_b)):
                common_c += c[0]
            if len(common_c) == len(ln_a) - 1:
                return common_c

    return None
    

def part_2b(input_str):
    for ln_a in input_str.splitlines():
        for ln_b in input_str.splitlines()[1:]:
            if len(ln_a) != len(ln_b):
                continue
            common_c = ""
            diff_count = 0
            for a,b in zip(ln_a, ln_b):
                if a != b :
                    diff_count += 1
                    if diff_count > 1 :
                        break
                else:
                    common_c += a
            
            if diff_count == 1:
                return common_c

    return None
                



main(sys.argv[1])